rtbkit/core/banker/go_account.cc

```diff
@@ -0,0 +1,262 @@
+    else imp = 0;
+
+    if (json.isMember("spend")) spend = json["spend"].asInt();
+    else spend = 0;
+}
+
+bool
+GoPostAuctionAccount::win(Amount winPrice)
+{
+    spend += winPrice.value;
+    imp += 1;
+    return true;
+}
+
+void
+GoPostAuctionAccount::toJson(Json::Value &account)
+{
+    account["imp"] = int64_t(imp);
+    account["spend"] = int64_t(spend);
+    GoBaseAccount::toJson(account);
+}
+
+//Account Base
+
+GoBaseAccount::GoBaseAccount(AccountKey &key)
+    : name(key.toString()), parent(key.parent().toString())
+{
+}
+
+GoBaseAccount::GoBaseAccount(Json::Value &json)
+{
+    if (json.isMember("name")) name = json["name"].asString();
+    else name = ""; // TODO: throw error instead?
+
+    if (json.isMember("parent")) parent = json["parent"].asString();
+    else parent = "";
+}
+
+void
+GoBaseAccount::toJson(Json::Value &account)
+{
+    account["name"] = name;
+    account["parent"] = parent;
+}
+
+// Accounts
+
+GoAccounts::GoAccounts() : accounts{}
+{
+}
+
+void
+GoAccounts::add(AccountKey &key, GoAccountType type)
+{
+    auto account = accounts.find(key);
+    if (account == accounts.end()) {
+        return;
+    }
+
+    std::lock_guard<std::mutex> guard(this->mutex);
+    accounts[key] = GoAccount(key, type);
+}
+
+void
+GoAccounts::addFromJsonString(std::string jsonAccount)
+{
+    Json::Value json = Json::parse(jsonAccount);
+    if (json.isMember("type") && json.isMember("name")) {
+        string name = json["name"].asString();
+
+        if (get(AccountKey(name))) return;
+
+        auto account = GoAccount(json);
+        {
+            std::lock_guard<std::mutex> guard(this->mutex);
+            accounts[name] = account;
+        }
+        //cout << "account in map: " << accounts[name].toJson() << endl;
+    } else {
+        cout << "error: type or name not parsed" << endl;
+    }
+}
+
+void
+GoAccounts::updateBalance(AccountKey &key, int64_t newBalance)
+{
+    auto account = get(key);
+
+    if (!account) return;
+    
+    std::lock_guard<std::mutex> guard(this->mutex);
+    account->router->balance = newBalance;
+}
+
+bool
+GoAccounts::bid(const AccountKey &key, Amount bidPrice)
+{
+    auto account = get(key);
+
+    if (!account) return false;
+
+    std::lock_guard<std::mutex> guard(this->mutex);
+    account->bid(bidPrice);
+    return false;
+}
+
+bool
+GoAccounts::win(const AccountKey &key, Amount winPrice)
+{
+    auto account = get(key);
+
+    if (!account) {
+        cout << "account not found, unaccounted win: " << key.toString()
+             << " " << winPrice.toString() << endl;
+        return false;
+    }
+
+    if (account->type != POST_AUCTION) {
+        throw ML::Exception("GoAccounts::win: attempt win on non POST_AUCTION account");
+    }
+
+    std::lock_guard<std::mutex> guard(this->mutex);
+    account->win(winPrice);
+    return true;
+}
+
+GoAccount*
+GoAccounts::get(const AccountKey &key)
+{
+    auto account = accounts.find(key);
```

Comment:
I guess you should lock here too but that will make the other lock fails... probably better to add a private function that doesn't lock and use that.